### PR TITLE
[Relay/TOPI][ONNX/TFLite] Refactor MATRIX_SET_DIAG Operator for Relay/TOPI to support ONNX Trilu operator

### PR DESCRIPTION
**Pull request by bfgoldstein**
This commit is based on PR #9329 proposed by @shengxinhu.

Refactor MATRIX_SET_DIAG operator in Relay/TOPI to support ONNX Trilu operator;
    + Fixed issues related to shape transformation of inputs in TFLite and ONNX frontend ops.

@AndrewZhaoLuo PTAL

### CODE REVIEW

File: src/relay/op/tensor/transform.cc [hunk 1]
```diff
@@ -3880,9 +3866,11 @@ RELAY_REGISTER_OP("matrix_set_diag")
         **sub_diag_right_align** Bool, true iff sub-diagonal is right aligned (left-padded).
     )code" TVM_ADD_FILELINE)
     .set_attrs_type<MatrixSetDiagAttrs>()
-    .set_num_inputs(2)
+    .set_num_inputs(4)
     .add_argument("input", "Tensor", "Input Tensor.")
     .add_argument("diagonal", "Tensor", "Values to be filled in the diagonal.")
+    .add_argument("k1", "Tensor", "ILower limit (included) of the range of diagonals.")
```
Comment (jwfromm) [inline src/relay/op/tensor/transform.cc:3872]: Small typo here "ILower" -> "Lower"

File: tests/python/topi/python/test_topi_transform.py [hunk 2]
```diff
@@ -752,21 +752,38 @@ def check_device(target, dev):
 def verify_matrix_set_diag(input_shape, diagonal_shape, dtype, k=0, align="RIGHT_LEFT"):
     input = te.placeholder(shape=input_shape, name="input", dtype=dtype)
     diagonal = te.placeholder(shape=diagonal_shape, name="diagonal", dtype=dtype)
-    matrix_set_diag_result = topi.transform.matrix_set_diag(input, diagonal, k, align)
+    k1 = te.placeholder(shape=(1,), name="k1", dtype="int64")
+    k2 = te.placeholder(shape=(1,), name="k2", dtype="int64")
+    matrix_set_diag_result = topi.transform.matrix_set_diag(input, diagonal, (k1, k2), align)
+
+    k_one, k_two = None, None
+    if isinstance(k, (tuple, list)):
```
Comment (jwfromm) [inline tests/python/topi/python/test_topi_transform.py:760]: Can you add some comments to this test? It's a little hard to follow.

**Review by jwfromm — COMMENTED** [review]
Thanks for this change, it overall seems excellent. I left a few nitpicks and questions.

File: tests/python/topi/python/test_topi_transform.py [hunk 3]
```diff
@@ -1240,9 +1257,6 @@ def test_sparse_to_dense():
 @tvm.testing.uses_gpu
 def test_matrix_set_diag():
     for dtype in ["float32", "int32"]:
-        verify_matrix_set_diag((2, 2), (2,), dtype)
-        verify_matrix_set_diag((4, 3, 3), (4, 3), dtype)
-        verify_matrix_set_diag((2, 3, 4), (2, 3), dtype, 1)
```
[the remark this answers is not in the document]
Comment (bfgoldstein) [reply tests/python/topi/python/test_topi_transform.py:1245]: This has been removed in the original PR. I ended up not putting it back and testing it again. Thanks for pointing it out since the test_matrix_set_diag is outputting an error for these three cases. I will check it out.

File: tests/python/topi/python/test_topi_transform.py (hunk 2 above)
Comment (bfgoldstein) [reply tests/python/topi/python/test_topi_transform.py:760]: Sure. I will add some comments and update the PR.

File: include/tvm/topi/transform.h [hunk 4]
```diff
@@ -1886,13 +1883,16 @@ inline Tensor matrix_set_diag(const Tensor& input, const Tensor& diagonal, int k
                                        : 0,
                 sub_diag_right_align ? get_offset(input->shape[ndim], input->shape[ndim - 1] + k)
                                      : 0);
-          }
+            return k;
+          };
+          auto get_k = [&]() { return if_then_else(k1(0) == k2(0), k1(0), multi_diagonals()); };
+          k = get_k();
```
Comment (AndrewZhaoLuo) [inline include/tvm/topi/transform.h:1889]: can you just do something like `k = if_then_else(k1(0) == k2(0), k1(0), multi_diagonals());`?

File: src/relay/op/tensor/transform.cc [hunk 5]
```diff
@@ -3812,59 +3812,45 @@ TVM_REGISTER_NODE_TYPE(MatrixSetDiagAttrs);
 bool MatrixSetDiagRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
                       const TypeReporter& reporter) {
   // `types` contains: [input, diagonal, result]
```
Comment (AndrewZhaoLuo) [inline src/relay/op/tensor/transform.cc:3814]: make sure to update this comment

File: include/tvm/topi/transform.h (hunk 4 above)
Comment (bfgoldstein) [reply include/tvm/topi/transform.h]: Thanks, Andrew. I have changed it.

File: src/relay/op/tensor/transform.cc (hunk 5 above)
Comment (bfgoldstein) [reply src/relay/op/tensor/transform.cc]: Done

**Comment by bfgoldstein** [comment]
> @bfgoldstein what is the status on this one?

@mikepapadim, it is on hold. I will try to work on that by early next week.

**Comment by mikepapadim** [comment]
> > @bfgoldstein what is the status on this one?
> 
> @mikepapadim, it is on hold. I will try to work on that by early next week.

no worries, I was looking into the `trilu` ticket and I bumped into this PR.

**Comment by bfgoldstein** [comment]
> > > @bfgoldstein what is the status on this one?
> > 
> > 
> > @mikepapadim, it is on hold. I will try to work on that by early next week.
> 
> no worries, I was looking into the `trilu` ticket and I bumped into this PR.

No problem! I bump into two special test cases which the TOPI is not passing. I had to put it on hold due to priorities.
